File: pick_helpers.py
#!/usr/bin/env python3
"""
pick_helpers.py — Shared Pick-Verarbeitung
============================================

Helper-Funktionen die von MEHREREN Modulen verwendet werden:
  · is_legitimate_pick(p)      — Single Source of Truth: zählt der Pick fürs Tracking?
  · pick_direction(p)          — Direction-Lookup (delegated to pick_constants)
  · picks_are_incompatible(...) — Cross-Market-Konflikt-Check
  · hero_sort_key(p)           — Sortierung: saferAlt > BET > Edge desc
  · select_hero(picks)         — Wählt den Hero aus einer Pick-Liste

Wird sowohl von Pick-Generierung als auch von Tracking + Renderern verwendet.
Backwards-compatible: ohne pick_constants läuft jeder Caller einfach mit
False/None Default → keine Regression möglich.
"""
from __future__ import annotations

try:
    from pick_constants import (
        DIRECTION_MAP,
        INCOMPATIBLE,
        are_directions_incompatible,
        get_pick_direction,
    )
except ImportError:
    # Soft-Fallback wenn pick_constants nicht importierbar (z.B. CI ohne PYTHONPATH)
    DIRECTION_MAP = {}
    INCOMPATIBLE = set()
    def are_directions_incompatible(d1, d2): return False
    def get_pick_direction(market): return None


# ─────────────────────────────────────────────────────────────────────────────
#  is_legitimate_pick — wird der Pick fürs Tracking gezählt?
# ─────────────────────────────────────────────────────────────────────────────
def is_legitimate_pick(p: dict | None) -> bool:
    """True wenn der Pick fürs Tracking/Display zählt.

    NICHT legitim:
      · trackingExcluded=True (vom Konflikt-Filter markiert)
      · None / leerer dict

    Ein Pick mit Result=WIN/LOSS/VOID ist trotzdem legitim — nur
    trackingExcluded macht ihn illegitim.
    """
    if p is None or not isinstance(p, dict):
        return False
    if p.get("trackingExcluded"):
        return False
    return True


def pick_direction(p: dict | None) -> str | None:
    """Returns Direction des Picks via Market-Lookup."""
    if not p or not isinstance(p, dict):
        return None
    return get_pick_direction(p.get("market"))


def picks_are_incompatible(p_a: dict, p_b: dict) -> bool:
    """True wenn die beiden Picks logisch unvereinbar als gleichzeitige BETs sind."""
    return are_directions_incompatible(pick_direction(p_a), pick_direction(p_b))


# ─────────────────────────────────────────────────────────────────────────────
#  hero_sort_key — Sortier-Logik für Hero-Auswahl
# ─────────────────────────────────────────────────────────────────────────────
def hero_sort_key(p: dict) -> tuple:
    """Sortierreihenfolge für Hero-Selection.

    Tuple-Vergleich: kleinerer Wert wird zuerst gewählt.
      1. saferAlt-Picks bevorzugen (False=0, True=1 → invertiert mit "0 if X else 1")
      2. BET vor ABWÄGEN
      3. Edge descending

    Beispiel:
        sorted(picks, key=hero_sort_key)[0]  → Hero
    """
    has_safer = bool(p.get("saferAltFor"))
    is_bet    = p.get("verdict") == "BET"
    edge_pp   = float(p.get("edgePP") or 0)
    return (
        0 if has_safer else 1,   # safer zuerst
        0 if is_bet    else 1,   # BET zuerst
        -edge_pp,                # höhere Edge zuerst
    )


def select_hero(picks: list, include_abwaegen: bool = True) -> dict | None:
    """Wählt den Hero aus einer Pick-Liste.

    Filtert auto:
      · trackingExcluded raus (is_legitimate_pick)
      · WATCH/STAT raus (nur BET + optional ABWÄGEN)

    Returns: Hero-Pick dict oder None wenn nichts qualifiziert.
    """
    valid_verdicts = {"BET", "ABWÄGEN"} if include_abwaegen else {"BET"}
    candidates = [
        p for p in picks
        if is_legitimate_pick(p) and p.get("verdict") in valid_verdicts
    ]
    if not candidates:
        return None
    return sorted(candidates, key=hero_sort_key)[0]


# ─────────────────────────────────────────────────────────────────────────────
#  Cross-Market-Konflikt-Filter
# ─────────────────────────────────────────────────────────────────────────────
def find_picks_conflicting_with_hero(picks: list, hero: dict) -> list[dict]:
    """Returns Liste von Picks die direktional mit Hero unvereinbar sind.

    Hilfreich für:
      · Renderer-Filter (Secondary-Picks ausblenden)
      · Tracker (als VOID/trackingExcluded markieren)
      · Validator (Errors melden)
    """
    if not hero:
        return []
    hero_dir = pick_direction(hero)
    if not hero_dir:
        return []
    conflicts = []
    for p in picks:
        if p is hero:
            continue
        if not is_legitimate_pick(p):
            continue
        d = pick_direction(p)
        if d and are_directions_incompatible(hero_dir, d):
            conflicts.append(p)
    return conflicts


if __name__ == "__main__":
    # Smoketest
    print("pick_helpers.py smoketest:")
    sample = [
        {"market": "Heimsieg",                "verdict": "BET",     "edgePP": 12, "odds": 2.01},
        {"market": "DNB: Heimteam",           "verdict": "ABWÄGEN", "edgePP": 6,  "odds": 1.46},
        {"market": "Über 1.5 Tore",           "verdict": "BET",     "edgePP": 9,  "odds": 1.40, "saferAltFor": "Über 3.5 Tore"},
        {"market": "AH Auswärts +0.5",        "verdict": "ABWÄGEN", "edgePP": 12, "odds": 1.87},  # inkompatibel mit Heimsieg!
    ]
    hero = select_hero(sample)
    print(f"  Hero: {hero.get('market')} edge={hero.get('edgePP')}pp (saferAltFor={hero.get('saferAltFor')!r})")
    conflicts = find_picks_conflicting_with_hero(sample, hero)
    print(f"  Konflikte: {[c.get('market') for c in conflicts]}")
    assert hero["market"] == "Über 1.5 Tore", "Hero sollte saferAlt-Pick sein"
    print("  ✅ Hero-Sort: saferAlt vor BET-Edge")

File: _pick_helpers.js
/*
 * _pick_helpers.js — JavaScript-Pendant zu pick_helpers.py
 * ─────────────────────────────────────────────────────────
 * Single Source of Truth für die JS-Seite:
 *   · DIRECTION_MAP   (29 Märkte → Richtung)
 *   · INCOMPATIBLE    (8 Konflikt-Paare)
 *   · isLegitimatePick(p)        — Tracker-Filter
 *   · arePicksConflicting(p1,p2) — Cross-Market-Check
 *   · findConflictingPicks(hero, others) — UI-Filter
 *   · heroSortKey(p)             — saferAlt > BET > Edge desc
 *
 * Werte spiegeln pick_constants.json (Python = Master).
 * Drift wird per Test (tests/test_js_pick_helpers.py) verboten.
 *
 * Lade-Pattern:
 *   Browser:  <script src="_pick_helpers.js"></script>  → window.CocoBetPicks
 *   Node:     const H = require('./_pick_helpers.js');
 */
(function (root, factory) {
  'use strict';
  if (typeof module === 'object' && module.exports) {
    module.exports = factory();
  } else {
    root.CocoBetPicks = factory();
  }
})(typeof self !== 'undefined' ? self : this, function () {
  'use strict';

  // ── DIRECTION_MAP ─────────────────────────────────────────────────────────
  // Mirror von pick_constants.json DIRECTION_MAP — IDENTISCH halten.
  // Schema: Market-Label → Richtungs-Bucket
  const DIRECTION_MAP = Object.freeze({
    'Heimsieg':                'homeStrong',
    'Auswärtssieg':            'awayStrong',
    'Unentschieden':           'drawOnly',
    'Doppelte Chance — 1X':    'homeBias',
    'Doppelte Chance — X2':    'awayBias',
    'Doppelte Chance — 12':    'decisive',
    'AH Heim −0.25':           'homeStrong',
    'AH Heim −0.5':            'homeStrong',
    'AH Heim −0.75':           'homeStrong',
    'AH Heim −1.0':            'homeStrong',
    'AH Heim −1.25':           'homeStrong',
    'AH Heim −1.5':            'homeStrong',
    'AH Heim −1.75':           'homeStrong',
    'AH Heim −2.0':            'homeStrong',
    'AH Heim −2.25':           'homeStrong',
    'AH Auswärts +0.25':       'awayStrong',
    'AH Auswärts +0.5':        'awayStrong',
    'AH Auswärts +0.75':       'awayStrong',
    'AH Auswärts +1.0':        'awayStrong',
    'AH Auswärts +1.25':       'awayStrong',
    'AH Auswärts +1.5':        'awayStrong',
    'AH Auswärts +1.75':       'awayStrong',
    'AH Auswärts +2.0':        'awayStrong',
    'AH Auswärts +2.25':       'awayStrong',
    'DNB: Heimteam':           'homeStrong',
    'DNB: Auswärtsteam':       'awayStrong',
    'Über 0.5 Tore':           'over',
    'Über 1.5 Tore':           'over',
    'Über 2.5 Tore':           'over',
    'Über 3.5 Tore':           'over',
    'Über 4.5 Tore':           'over',
    'Unter 0.5 Tore':          'under',
    'Unter 1.5 Tore':          'under',
    'Unter 2.5 Tore':          'under',
    'Unter 3.5 Tore':          'under',
    'Unter 4.5 Tore':          'under',
    'Beide Teams treffen':     'over',
    'Beide Teams treffen — Ja': 'over',
    'Beide Teams treffen — Nein': 'under'
  });

  // ── INCOMPATIBLE pairs ────────────────────────────────────────────────────
  // Set von "dir1|dir2" Strings — symmetrisch (beide Richtungen drin)
  const INCOMPATIBLE = Object.freeze(new Set([
    'homeStrong|awayStrong', 'awayStrong|homeStrong',
    'homeStrong|awayBias',   'awayBias|homeStrong',
    'homeStrong|drawOnly',   'drawOnly|homeStrong',
    'homeBias|awayStrong',   'awayStrong|homeBias',
    'awayStrong|drawOnly',   'drawOnly|awayStrong',
    'decisive|drawOnly',     'drawOnly|decisive',
    'over|under',            'under|over'
  ]));

  // ── Public API ────────────────────────────────────────────────────────────

  /** Direction-Bucket für ein Market-Label. null wenn unbekannt. */
  function getPickDirection(market) {
    if (typeof market !== 'string') return null;
    return DIRECTION_MAP[market] || null;
  }

  /** True wenn zwei Direction-Buckets nicht koexistieren können. */
  function areDirectionsIncompatible(d1, d2) {
    if (!d1 || !d2) return false;
    return INCOMPATIBLE.has(d1 + '|' + d2);
  }

  /**
   * True wenn ein Pick *legitim* für UI/Tracker/Stats ist.
   * Filtert trackingExcluded raus — sonst tauchen Konflikt-Picks
   * (z.B. AH Heim −0.5 wenn Hero X2 ist) wieder in der Card auf.
   * Spiegelt pick_helpers.is_legitimate_pick.
   */
  function isLegitimatePick(p) {
    if (p == null) return false;
    if (typeof p !== 'object') return true;
    return !p.trackingExcluded;
  }

  /** True wenn zwei Picks (per Market-Direction) miteinander kollidieren. */
  function arePicksConflicting(p1, p2) {
    if (!p1 || !p2) return false;
    const d1 = getPickDirection(p1.market);
    const d2 = getPickDirection(p2.market);
    return areDirectionsIncompatible(d1, d2);
  }

  /**
   * Filter-Helper: gibt aus `others` nur die Picks zurück,
   * die mit Hero-Pick KOLLIDIEREN. Nützlich um UI-Konflikte zu erkennen.
   */
  function findConflictingPicks(hero, others) {
    if (!hero || !Array.isArray(others)) return [];
    return others.filter(p => arePicksConflicting(hero, p));
  }

  /**
   * Hero-Sort-Key: Picks für Card-Auswahl sortieren.
   * Spiegelt pick_helpers.hero_sort_key (Python).
   *
   * Reihenfolge:
   *   1. saferAlt-Picks gewinnen (sicherere Quote schlägt riskante BET)
   *   2. dann BET vor ABWÄGEN vor SKIP
   *   3. dann höchste Edge zuerst
   *
   * Verwendung:  picks.sort((a, b) => heroSortKey(a) - heroSortKey(b))  ?
   * → Liefert STATTDESSEN ein Tuple-Array zurück, damit man stable sort hat.
   *   Nutzung: picks.sort((a, b) => {
   *     const ka = heroSortKey(a), kb = heroSortKey(b);
   *     for (let i = 0; i < ka.length; i++) {
   *       if (ka[i] < kb[i]) return -1;
   *       if (ka[i] > kb[i]) return 1;
   *     }
   *     return 0;
   *   });
   *
   * Komfort: heroSortCompare(a, b) direkt nutzen.
   */
  function heroSortKey(p) {
    if (!p) return [9, 9, 0];
    const saferRank  = p.saferAltFor ? 0 : 1;  // saferAlt zuerst
    const verdictRank = (p.verdict === 'BET') ? 0
                     : (p.verdict === 'ABWÄGEN') ? 1
                     : 2;
    const edgeNeg = -(Number(p.edgePP) || 0);  // höhere Edge zuerst
    return [saferRank, verdictRank, edgeNeg];
  }

  /** Komfort-Wrapper: direkt nutzbar in Array.sort(). */
  function heroSortCompare(a, b) {
    const ka = heroSortKey(a);
    const kb = heroSortKey(b);
    for (let i = 0; i < ka.length; i++) {
      if (ka[i] < kb[i]) return -1;
      if (ka[i] > kb[i]) return  1;
    }
    return 0;
  }

  /**
   * Wählt den Hero-Pick aus einer Liste.
   * Nur legitime Picks werden berücksichtigt.
   * Gibt null zurück wenn keine vorhanden.
   */
  function selectHero(picks) {
    if (!Array.isArray(picks)) return null;
    const legitimate = picks.filter(isLegitimatePick);
    if (legitimate.length === 0) return null;
    return legitimate.slice().sort(heroSortCompare)[0];
  }

  return Object.freeze({
    DIRECTION_MAP,
    INCOMPATIBLE,
    getPickDirection,
    areDirectionsIncompatible,
    isLegitimatePick,
    arePicksConflicting,
    findConflictingPicks,
    heroSortKey,
    heroSortCompare,
    selectHero,
    _version: '1.0.0'
  });
});
